File: dataStructure/数组/MyArray.java
// TODO 实现数组类
public class MyArray<E> {
    private E[] data;
    private int size;

    /**
     *
     * @param capacity 容量
     */
    public Array(int capacity) {
        data = (E[])new Object[capacity];
        size = 0
    }

    // 无参数构造函数，不传的话，默认是10
    public Array() {
        this(10)
    }

    // 获取数组元素个数
    public int getSize() {
        return size;
    }

    // 获取数组容量
    public int getCapacity() {
        return data.length;
    }

    public boolean isEmpty() {
        return size === 0
    }

    // 向数组最后添加一个元素
    public void addLast(E e) {
        add(size, e);
    }

    public void addFirst(E e) {
        add(0, e);
    }

    // 插入数据
    public void add(int index, E e) {
        if(index < 0 || index > size) throw new IllegalArgumentException('index is invalid');
        if(size == data.length) {
            // TODO 动态数组
            resize(2 * data.length)
        }
        for (int i = size - 1; i >= index; i--) data[i + 1] = data[i];
        data[index] = e;
        size++;
    }

    public E get(int idx) {
        if(idx < 0 || idx >= size) throw new IllegalArgumentException('get failed');
        return data[idx];
    }

    public void set(int idx, E e) {
        if(idx < 0 || idx >= size) throw new IllegalArgumentException('set failed');
        data[idx] = e;
    }

    public boolean contains(E e) {
        for (int i = 0; i < size; i++) {
            if(data[i].equals(e)) return true;
        }
        return false;
    }

    public int find(E e) {
        for (int i = 0; i < size; i++) {
            if(data[i].equals(e)) return i;
        }
        return -1;
    }

    public E remove(int idx) {
        if(idx < 0 || idx >= size) throw new IllegalArgumentException('remove failed');
        int ret = data[idx];
        for (int i = idx + 1; i < size; i++) data[i - 1] = data[i];
        size--;
        data[size] = null;
        if(size == data.length / 2) resize(data.length / 2); // 缩小数组
        return ret;
    }

    public E removeFirst() {
        return remove(0);
    }

    public E removeLast() {
        return remove(size - 1);
    }

    public void removeElement(E e) {
        int idx = find(e);
        if(idx != -1) remove(idx);
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        res.append(String.format("Array: size = %d, capacity = %d\n", size, data.length));
        res.append('[');
        for (int i = 0; i < size; i++) {
            res.append(data[i]);
            if(i !=  size - 1) res.append(',');
        }
        res.append(']');
        return res.toString();
    }

    private void resize(int newCapacity) {
        E[] newData = (E[]) new Object[newCapacity];
        for (int i = 0; i < size; i++) {
            newData[i] = data[i]
        }
        data = newData
    }

}

File: dataStructure/数组/MyArray.ts
class MyArray<E> {
    private data: E[] = []
    private size: number = 0
    constructor(capacity?: number) {
        this.data = new Array<E>(capacity ? capacity : 10)
    }
    public getSize(): number {
        return this.size
    }

    public getCapacity(): number {
        return this.data.length
    }
    public isEmpty(): boolean {
        return this.size === 0
    }

    public addLast(e: E): void {
        this.add(this.size, e)
    }

    public addFirst(e:E): void {
        this.add(0, e)
    }

    // 插入数据
    public add(index: number, e: E): void {
        if(index < 0 || index > this.size) throw new Error('index is invalid')
        if(this.size === this.data.length) {
            // TODO 动态数组
            this.resize(2 * this.data.length)
        }
        for (let i = this.size - 1; i >= index; i--) this.data[i + 1] = this.data[i]
        this.data[index] = e
        this.size++
    }

    public get(idx: number): E {
        if(idx < 0 || idx >= this.size) throw new Error('get failed')
        return this.data[idx]
    }

    public set(idx: number, e: E) {
        if(idx < 0 || idx >= this.size) throw new Error('set failed')
        this.data[idx] = e
    }

    public contains(e: E):boolean {
        for (let i = 0; i < this.size; i++) {
            if(this.data[i] == e) return true;
        }
        return false;
    }

    public remove(idx: number): E {
        if(idx < 0 || idx >= this.size) throw new Error('remove failed')
        let ret = this.data[idx];
        for (let i = idx + 1; i < this.size; i++) this.data[i - 1] = this.data[i];
        this.size--;
        if(this.size === this.data.length / 2) this.resize(this.data.length / 2); // 缩小数组
        return ret
    }

    public removeFirst(): E {
        return this.remove(0 )
    }

    public removeLast(): E {
        return this.remove(this.size - 1)
    }

    public find(e: E): number {
        for (let i = 0; i < this.size; i++) {
            if(this.data[i] == e) return i
        }
        return -1;
    }

    public removeElement(e: E): void {
        let idx = this.find(e);
        if(idx !== -1) this.remove(idx);
    }

    public toString(): string {
        let str: string = `Array: size = ${this.size}, capacity = ${this.data.length}\n`
        str+='['
        for (let i = 0; i < this.size; i++) {
            str+=this.data[i];
            if(i != this.size - 1) str+=',';
        }
        str+=']'
        return str
    }

    // 扩大数组大小
    private resize(newCapacity: number): void {
        let newData: E[] = new Array<E>(newCapacity);
        for (let i = 0; i < this.size; i++) {
            newData[i] = this.data[i]
        }
        this.data = newData
    }

}


// 测试使用代码
let myArr = new MyArray<number>(20)
for (let index = 0; index < 10; index++) {
    myArr.addLast(index)
}
myArr.add(1, 100)
console.log(myArr.toString());
